test/jdk/java/util/UUID/UUIDTest.java

```diff
@@ -49,56 +57,76 @@ public static void main(String[] args) throws Exception {
         compareTo();
     }
 
-    // Verify that list.contains detects UUID collisons
-    private static void containsTest() throws Exception {
-        List list = new LinkedList();
-        list.add(new UUID(4,4));
-        if (!list.contains(new UUID(4,4)))
-            throw new Exception("contains test did not work as expected");
+    private static void negativeTest() throws Exception {
+        Set<UUID> set = new HashSet<>();
+        set.add(new UUID(4, 4));
+        if (set.add(new UUID(4, 4))) {
+            throw new Exception("Contains test does not work as expected");
+        }
     }
 
     private static void randomUUIDTest() throws Exception {
-        List list = new LinkedList();
-        for (int i=0; i<100; i++) {
-            UUID u1 = UUID.randomUUID();
-            if (4 != u1.version()) {
-                throw new Exception("bad version");
+        Set<UUID> set = new HashSet<>();
+        for (int i = 0; i < COUNT; i++) {
+            UUID u = UUID.randomUUID();
+            if (u.version() != 4) {
+                throw new Exception("Bad version: " + u);
+            }
+            if (u.variant() != 2) {
+                throw new Exception("Bad variant: " + u);
+            }
+            if (!set.add(u)) {
+                throw new Exception("UUID collision: " + u);
```

Comment:
I would be concerned that if this failure was reported, it would be intermittent, hard to track down, and not reproducible.  Without a hook for the generator and the seed, its just going to be noise in the testing.

Reply:
This is a non-practical concern, IMO. By spec, `UUID.randomUUID` is generated from the cryptographically secure random, with >120 bits of randomness, so the collision is extremely unlikely. Collision math involves birthday paradox, but Wikipedia article on UUIDs fortunately gives us the approximated solutions already:  https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions 

Quote: "Thus, the probability to find a duplicate within 103 trillion version-4 UUIDs is one in a billion." 

In other words, finding a collision in this test with 1M UUIDs points to the implementation issue, not a test bug, with a very high probability. In yet another words, if a unit test with 1M UUIDs is able to find a collision, then this is a strong signal that many production systems that assume extremely low collision probability are up for subtle misbehavior.

Reply:
My point was that its probably not practical to test (more than once).  
If it fails, it will be considered just as you propose and disregarded and in the meantime consumes test cycles in each of the test contexts. Either provide more information about the conditions under which it failed or remove it.

Reply:
Sorry, I have trouble following the argument here.

Let me re-iterate: the probability for bona-fide collision is so vanishingly low, the test failure here is a strong signal that something is wrong with the implementation. We can put more guidance in the test comments there, like "This is extremely unlikely to happen. If you see this failing, this highly likely points to the implementation bug, rather than the odd chance."

What I expect to happen when that test fails, is that it prompts the investigation with multiple stress tests to get a better estimate of the actual collision rate. Assuming we actually see a collision, it is likely to be caused by much higher probability error somewhere in the code. In fact, if this test is _actually noisy_ to the point it becomes a testing problem, this already gives us the signal that actual collision rate is many orders of magnitude higher than math predicts, and this becomes an even _stronger_ signal that random UUIDs are seriously broken for practical use.

Reply:
Two thoughts here. 
The random number source (SecureRandom) should have its own tests, UUID has a simple dependency on a generator.  
The test for UUID is that it composes the bits into the UUID correctly.  The randomness of the generator should be factored out.

Second, to raise concern about collisions, then the test should not throw on the first detected collision but complete the cycle and provide the simple stats on the number of collisions per COUNT (1,000,000).

All those 5 second test runs add up.

Reply:
> Two thoughts here. The random number source (SecureRandom) should have its own tests, UUID has a simple dependency on a generator. The test for UUID is that it composes the bits into the UUID correctly. The randomness of the generator should be factored out.

This test verifies the way UUID uses the SecureRandom. Think about this as the less of a unit, and more of the integration test. This would be even more important once things like https://bugs.openjdk.org/browse/JDK-8308804 show up.

> Second, to raise concern about collisions, then the test should not throw on the first detected collision but complete the cycle and provide the simple stats on the number of collisions per COUNT (1,000,000).

All right, that we can do. See new commit.

> All those 5 second test runs add up.

Yes, and it would be sad to waste those 5 seconds on something irrelevant. I would argue that `UUID.randomUUID` breakage would be very unfortunate for the real world systems. Spending 5 seconds per test run on it is a good investment here.

Reply:
We're into the area of diminishing returns.  
I'd expect the implementation bug mentioned above would be found by far fewer iterations. 
Thanks for the test change.